crates/volta-core/src/session.rs

```diff
@@ -108,10 +108,13 @@ impl Session {
         self.project.get()
     }
 
-    pub fn current_platform(&self) -> Fallible<Option<Rc<PlatformSpec>>> {
+    pub fn current_platform(&self) -> Fallible<Option<SourcedPlatformSpec>> {
         match self.project_platform()? {
-            Some(platform) => Ok(Some(platform)),
-            None => self.user_platform(),
+            Some(platform) => Ok(Some(SourcedPlatformSpec::project(platform))),
+            None => {
```

Comment:
*Certainly* not a blocker, but you can simplify this a bit:

```rust
Ok(user.map(SourcedPlatformSpec::user))
```

You can also reduce it into a single expression if you so desire:

```rust
None => self
    .user_platform()
    .map(|platform| platform.map(SourcedPlatformSpec::user)),
```

The form I actually find *clearest*, though, is reworking the whole thing to make the decision tree clearer:

```rust
match (self.project_platform()?, self.user_platform()?) {
    (Some(platform), _) => Ok(Some(SourcedPlatformSpec::project(platform))),
    (_, Some(platform)) => Ok(Some(SourcedPlatformSpec::user(platform))),
    _ => Ok(None),
}
```

Reply:
I did consider the last of those, and I agree it makes the decision tree clearer, the issue I had with it is that it means the user platform file needs to be read and parsed for every invocation of a tool, even if everything is specified in the Project settings. With the current approach as well as the old code, the user platform isn't looked at unless it's needed.

This may be an early optimization, but especially when adding the verbose messaging, keeping the file access the same seemed important.

I also find the single-expression form slightly harder to grok, with a map within a map, but I like the improvement of using point-free style for the map function.

Reply:
Totally makes sense. The fact that it's doing a file lookup *does* make a difference here, and I'm inclined to agree with you.

Another alternative I considered here, which is the longest but may be the next-best option to preserve that *and* keep the decision tree equally clear:

```rust
if let Some(platform) = self.project_platform()? {
    Ok(Some(SourcedPlatformSpec::project(platform)))
} else if let Some(platform) = self.user_platform()? {
    Ok(Some(SourcedPlatformSpec::user(platform)))
} else {
    Ok(None)
}
```

At the end of the day, any of these are fine, but I did find the flow a little hard to follow in the original `match`, which is why I mentioned it!

Reply:
That’s a good idea, and makes the changes in the follow-on PR (#470) a little more explicit, so I think I’ll look into that! Thanks!